File: backend/app/models/interview_scheduling.py
# backend/models/interview_scheduling.py
from datetime import datetime
import uuid
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from . import job_posting
from app import db

class InterviewSchedule(db.Model):
    __tablename__ = 'interview_schedules'
    
    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    organization_id = db.Column(db.String(36), db.ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    
    # Informations sur le recruteur
    recruiter_id = db.Column(db.String(36), db.ForeignKey("users.id"), nullable=False)
    
    # Informations sur le candidat
    candidate_name = db.Column(db.String(255), nullable=False)
    candidate_email = db.Column(db.String(255), nullable=False)
    candidate_phone = db.Column(db.String(50), nullable=True)
    
    # Détails de l'entretien
    title = db.Column(db.String(255), nullable=False)
    description = db.Column(db.Text, nullable=True)
    position = db.Column(db.String(255), nullable=True)
    scheduled_at = db.Column(db.DateTime, nullable=False)
    duration_minutes = db.Column(db.Integer, default=30)
    timezone = db.Column(db.String(50), default="Africa/Douala")
    
    # Mode d'entretien
    mode = db.Column(db.Enum('autonomous', 'collaborative', name='interview_mode'), nullable=False)
    
    # Si un assistant IA spécifique est sélectionné
    ai_assistant_id = db.Column(db.String(36), db.ForeignKey("ai_assistants.id"), nullable=True)
    
    # Questions prédéfinies (stockées en JSON)
    predefined_questions = db.Column(db.JSON, nullable=True)
    
    # Statut et suivi
    status = db.Column(db.Enum(
        'scheduled', 'confirmed', 'in_progress', 'completed', 'canceled', 'no_show',
        name='interview_status'), default='scheduled')
    reminder_sent = db.Column(db.Boolean, default=False)
    cancellation_reason = db.Column(db.Text, nullable=True)
    
    # Lien de l'entretien (accessible par le candidat)
    access_token = db.Column(db.String(100), unique=True, nullable=False, default=lambda: str(uuid.uuid4()))
    
    # Résultat de l'entretien (lien vers l'entretien réalisé)
    interview_id = db.Column(db.String(36), db.ForeignKey("interviews.id"), nullable=True)
    
    # Métadonnées
    created_at = db.Column(db.DateTime, default=func.now())
    updated_at = db.Column(db.DateTime, default=func.now(), onupdate=func.now())
    
    # Relations
    organization = relationship("Organization")
    recruiter = relationship("User", foreign_keys=[recruiter_id])
    ai_assistant = relationship("AIAssistant")
    interview = relationship("Interview")
    # Ajouter ce champ à models/interview_scheduling.py dans la classe InterviewSchedule
    job_posting_id = db.Column(db.String(36), db.ForeignKey("job_postings.id"), nullable=True)
    job_posting = relationship("JobPosting", back_populates="interview_schedules")
    
    def __repr__(self):
        return f"<InterviewSchedule {self.candidate_name} - {self.scheduled_at}>"
    
    def to_dict(self):
        """Convertit l'objet en dictionnaire pour l'API"""
        return {
            'id': self.id,
            'organization_id': self.organization_id,
            'recruiter_id': self.recruiter_id,
            'recruiter_name': self.recruiter.name if self.recruiter else None,
            'candidate_name': self.candidate_name,
            'candidate_email': self.candidate_email,
            'candidate_phone': self.candidate_phone,
            'title': self.title,
            'description': self.description,
            'position': self.position,
            'scheduled_at': self.scheduled_at.isoformat() if self.scheduled_at else None,
            'duration_minutes': self.duration_minutes,
            'timezone': self.timezone,
            'mode': self.mode,
            'ai_assistant_id': self.ai_assistant_id,
            'ai_assistant_name': self.ai_assistant.name if self.ai_assistant else None,
            'predefined_questions': self.predefined_questions,
            'status': self.status,
            'reminder_sent': self.reminder_sent,
            'cancellation_reason': self.cancellation_reason,
            'access_token': self.access_token,
            'interview_id': self.interview_id,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
        }

File: tsx/src/types/interview-scheduling.ts
// frontend/types/interview-scheduling.ts

import { AvatarStatus } from "./avatar";

// Types pour les statuts d'entretien
export type ScheduleStatus = 'scheduled' | 'confirmed' | 'in_progress' | 'completed' | 'canceled' | 'no_show';

// Types pour les modes d'entretien - CORRIGÉ
export type InterviewMode = 'collaborative' | 'autonomous';

// Interface principale pour une planification d'entretien
export interface InterviewSchedule {
  id: string;
  candidate_name: string;
  candidate_email: string;
  candidate_phone?: string;
  title: string;
  description?: string;
  position: string;
  scheduled_at: string;
  duration_minutes: number;
  timezone: string;
  mode: InterviewMode; // collaborative ou autonomous
  ai_assistant_id?: string;
  predefined_questions?: string[];
  access_token: string;
  status: ScheduleStatus;
  interview_id?: string;
  reminder_sent?: boolean;
  organization_id: string;
  recruiter_id: string;
  
  was_confirmed_by_candidate: boolean;
  was_canceled_by_candidate: boolean;
  candidate_response_date?: string; // ISO date string
  cancellation_reason?: string;
  
  // Capacité de réponse candidat
  can_candidate_respond?: CandidateResponseCapability;
  
  // URLs de réponse (générées côté backend)
  candidate_response_urls?: CandidateResponseUrls;
  
  // Métadonnées audit
  created_at: string;
  updated_at: string;

  recruiter?: {
    id: string;
    name: string;
    email: string;
  };
  organization?: {
    id: string;
    name: string;
  };

  avatar?: AvatarStatus;

}

export interface CandidateResponseCapability {
  can_respond: boolean;
  reason: string; // Raison si le candidat ne peut pas répondre
  expires_at?: string; // Quand la capacité de réponse expire
}

// URLs de réponse candidat (générées par le backend)
export interface CandidateResponseUrls {
  confirm_url: string;
  cancel_url: string;
}

// Statut enrichi de réponse candidat
export interface CandidateResponseInfo {
  status: 'confirmed' | 'canceled' | 'pending' | 'expired';
  has_responded: boolean;
  response_date?: Date;
  cancellation_reason?: string;
  can_still_respond: boolean;
  time_until_expiry?: string;
  action_urls?: CandidateResponseUrls;
}

// Interface pour les données publiques (candidat)
export interface PublicScheduleData {
  id: string;
  title: string;
  description?: string;
  position: string;
  candidate_name: string;
  scheduled_at: string;
  duration_minutes: number;
  timezone: string;
  recruiter_name?: string;
  organization_name?: string;
  access_token: string;
  mode: InterviewMode;
  was_confirmed_by_candidate: boolean;
  was_canceled_by_candidate:boolean;
}

// Interface pour les filtres de recherche
export interface ScheduleFilters {
  status?: string;
  from_date?: string;
  to_date?: string;
  limit?: number;
  offset?: number;
}

// Interface pour la création/modification d'un entretien
export interface InterviewScheduleFormData {
  candidate_name: string;
  candidate_email: string;
  candidate_phone?: string;
  title: string;
  description?: string;
  position: string;
  scheduled_at: string;
  duration_minutes: number;
  timezone: string;
  mode: InterviewMode; // collaborative ou autonomous
  ai_assistant_id?: string;
  predefined_questions?: string[];
  job_id?: string
}

export interface InterviewScheduleUpdateData {
  scheduled_at?: string;
  duration_minutes?: number;
  timezone?: string;
  mode?: "autonomous" | "collaborative";
  ai_assistant_id?: string;
  predefined_questions?: string[];
}

// Interface pour les erreurs de validation
export interface ScheduleValidationErrors {
  [key: string]: string;
}

// Interface pour les options de mode d'entretien
export interface InterviewModeOption {
  value: InterviewMode;
  label: string;
  description: string;
  icon: string;
}

// Constantes pour les modes d'entretien
export const INTERVIEW_MODES: InterviewModeOption[] = [
  {
    value: 'collaborative',
    label: 'Collaboratif',
    description: 'Entretien avec assistant IA et interaction directe avec le recruteur',
    icon: 'users'
  },
  {
    value: 'autonomous',
    label: 'Autonome',
    description: 'Entretien autonome géré entièrement par l\'assistant IA',
    icon: 'bot'
  }
];

// Constantes pour les statuts
export const SCHEDULE_STATUSES: { value: ScheduleStatus; label: string; color: string }[] = [
  { value: 'scheduled', label: 'Planifié', color: '#3498db' },
  { value: 'confirmed', label: 'Confirmé', color: '#27ae60' },
  { value: 'in_progress', label: 'En cours', color: '#f39c12' },
  { value: 'completed', label: 'Terminé', color: '#2ecc71' },
  { value: 'canceled', label: 'Annulé', color: '#e74c3c' },
  { value: 'no_show', label: 'Absence candidat', color: '#95a5a6' }
];

// Types pour les anciennes interfaces (compatibilité)
export type InterviewStatus = 'scheduled' | 'in_progress' | 'completed' | 'cancelled';

export interface Interview {
  id: string;
  candidate_name: string;
  job_role: string;
  status: InterviewStatus;
  scheduled_time?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  score?: number;
}

export interface GetInterviewsFilters {
  status?: string;
  from_date?: string;
  to_date?: string;
}

export interface CreateInterviewScheduleRequest {
  candidate_name: string;
  candidate_email: string;
  candidate_phone?: string;
  title: string;
  description?: string;
  position: string;
  scheduled_at: string; // ISO date string
  duration_minutes?: number;
  timezone?: string;
  mode: 'autonomous' | 'collaborative';
  ai_assistant_id?: string;
  predefined_questions?: string[];
}

// Requête de mise à jour d'entretien
export interface UpdateInterviewScheduleRequest {
  candidate_name?: string;
  candidate_email?: string;
  candidate_phone?: string;
  title?: string;
  description?: string;
  position?: string;
  scheduled_at?: string;
  duration_minutes?: number;
  timezone?: string;
  mode?: 'autonomous' | 'collaborative';
  ai_assistant_id?: string;
  predefined_questions?: string[];
}

// Réponse API pour le statut candidat
export interface CandidateResponseStatusResponse {
  status: string;
  data: {
    current_status: string;
    can_respond: boolean;
    response_reason: string;
    confirmed_by_candidate: boolean;
    canceled_by_candidate: boolean;
    cancellation_reason?: string;
    response_date?: string;
    time_until_interview?: number; // en secondes
  };
}

// Réponse API pour l'aperçu email
export interface EmailPreviewResponse {
  status: string;
  data: {
    html_content: string;
    text_content: string;
    subject: string;
    recipient: string;
    contains_response_buttons: boolean;
  };
}

// Réponse API pour le renvoi d'invitation
export interface ResendInvitationResponse {
  status: string;
  message: string;
  data: {
    email_sent: boolean;
    sent_at: string;
    recipient: string;
  };
}

export interface CandidateResponseStatusProps {
  schedule: InterviewSchedule;
  onStatusUpdate?: (schedule: InterviewSchedule) => void;
  showActions?: boolean;
  compact?: boolean;
  autoRefresh?: boolean;
}

// Props pour le modal d'aperçu email
export interface EmailPreviewModalProps {
  visible: boolean;
  htmlContent: string;
  textContent?: string;
  onClose: () => void;
  candidateName: string;
  interviewTitle: string;
  loading?: boolean;
}

// État pour le hook de réponse candidat
export interface CandidateResponseHookState {
  loading: boolean;
  schedule: InterviewSchedule | null;
  error: string | null;
  lastUpdated: Date | null;
}

// Actions pour le hook de réponse candidat
export interface CandidateResponseHookActions {
  refreshStatus: () => Promise<void>;
  resendInvitation: () => Promise<boolean>;
  getEmailPreview: () => Promise<string>;
  canCandidateRespond: boolean;
  responseStatus: 'confirmed' | 'canceled' | 'pending' | 'expired';
  responseMetadata: CandidateResponseMetadata;
}

// Métadonnées de réponse candidat
export interface CandidateResponseMetadata {
  hasResponded: boolean;
  responseDate?: Date;
  cancellationReason?: string;
  canStillRespond: boolean;
  responseTimeLeft?: string;
  interviewInHours?: number;
}

// ====================================================================
// TYPES POUR LA LISTE ET FILTRES
// ====================================================================

// Filtre pour la liste des entretiens
export interface InterviewScheduleFilter {
  status?: string[];
  candidate_response?: 'confirmed' | 'canceled' | 'pending' | 'any';
  date_range?: {
    start: string;
    end: string;
  };
  search?: string;
  recruiter_id?: string;
  mode?: ('autonomous' | 'collaborative')[];
}

// Item de liste avec statut enrichi
export interface InterviewScheduleListItem extends InterviewSchedule {
  candidate_response_summary: {
    status_badge: string;
    status_color: 'success' | 'error' | 'processing' | 'default';
    quick_info: string;
    needs_attention: boolean;
  };
}

// ====================================================================
// TYPES POUR LES NOTIFICATIONS
// ====================================================================

// Notification de réponse candidat
export interface CandidateResponseNotification {
  id: string;
  type: 'candidate_confirmed' | 'candidate_canceled' | 'candidate_reminder';
  schedule_id: string;
  candidate_name: string;
  message: string;
  created_at: string;
  read: boolean;
  action_url?: string;
}

export enum Interview_Status {
  SCHEDULED = 'scheduled',
  CONFIRMED = 'confirmed',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  CANCELED = 'canceled',
  MISSED = 'missed'
}

export enum CandidateResponseStatus {
  CONFIRMED = 'confirmed',
  CANCELED = 'canceled', 
  PENDING = 'pending',
  EXPIRED = 'expired'
}

export enum Interview_Mode {
  AUTONOMOUS = 'autonomous',
  COLLABORATIVE = 'collaborative'
}

//
export interface ErrorDetails {
  type: 'validation' | 'duplicate' | 'network' | 'server' | 'unknown';
  message: string;
  details?: any;
  actionable?: boolean;
  suggestions?: string[];
}